### PR TITLE
[Need to discuss] New style of mergejson.py

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
The current flow of `data2json.sh`:

1. Creates text files, e.g. feats.scp, text, ...
1. Change each files to JSON files using `scp2json.py`
1. Merge JSON files into one

The second step looks redundant. So I modified it to

1. Creates text files, e.g. feats.scp, text, ...
1. Merge text files into a JSON file

For this, I created a new util: `merge_scp2json.py`

```sh
merge_scp2json.py ¥
  --input-scps feat:feats.scp shape:shape.scp:shape ¥
  --input-scps feat:feats2.scp shape:shape2.scp:shape ¥
  --output-scps text:text ¥
  --scps utt2spk 
```

The format of each arguments is `<key>:<filepath>:<type>`, where `<type>` indicated the data type registered in the output json, e.g. :

```sh
% cat foo.txt
a 3
b 10
% merge_scp2json.py --input-scps bar:foo.txt:int
{
    "utts": {
        "a": {
            "input": [
                {
                    "bar": 3,
                    "name": "input1"
                }
            ],
            "output": []
        },
        "b": {
            "input": [
                {
                    "bar": 10,
                    "name": "input1"
                }
            ],
            "output": []
        }
    }
}
```

### CODE REVIEW

**Comment by kamo-naoyuki** [comment]
Note that `shape ` of `shape:shape.txt:shape` is a special type:

```python
>>> shape('3,4')
[3, 4]
```

The benefit of this style, `text` files to a JSON files, is memory usage. 
JSON format cannot be load per elements, so we need to read JSON text fully at the first time, and it takes much memory if there are large samples (#533).
I designed `merge_scp2json.py` as reading each text files line by line and write each elements one by one.

**Comment by sw005320** [comment]
> 2. Change each files to JSON files using `scp2json.py`

I like your idea and please proceed with it. The reason I made such redundant process is that I was thinking of handling "3. Merge JSON files into one" with some command line JSON tools, e.g., `jq`. But it turns out that it cannot handle a complicated process. Also such command line tools are not so popular and writing our own merge scripts is easier. Now we also have a memory issue. We do not have to stick to my wrong design.

**Comment by kamo-naoyuki** [comment]
mergejson.py creates the intersection among the input json files at the first, but my tool assumes the keys of each input files are same and sorted and if some keys are different, the processing will stop in the middle with an error message. This is inevitable to read text files line by line.

So, we need to format text files in the preprocessing. e.g. `fix_data_dir.sh`
Is it o.k?

**Comment by sw005320** [comment]
> So, we need to format text files in the preprocessing. e.g. `fix_data_dir.sh`
Is it o.k?

Sure.

File: utils/merge_scp2json.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import codecs
+from io import open
+import json
+import logging
+import sys
+
+from espnet.utils.cli_utils import get_commandline_args
+
+PY2 = sys.version_info[0] == 2
+sys.stdin = codecs.getwriter('utf-8')(sys.stdin if PY2 else sys.stdin.buffer)
```
Comment (gtache) [inline utils/merge_scp2json.py]: It should be codecs.getreader for stdin

Comment (kamo-naoyuki) [reply utils/merge_scp2json.py]: @gtache Thanks, I fixed.

**Comment by kamo-naoyuki** [comment]
Note that this modification solves the memory issue for data2json.sh, but originally JSON causes memory issue if the items are so many because it's not supporting lazy loading.

HDF5 is a better choice as far as memory usage, but probably it degrades the readability because it's a binary format, simply. I don't have good idea.

**Comment by sw005320** [comment]
I guess it affects to the WSJ 2mix and speech translation recipes (https://github.com/espnet/espnet/search?q=mergejson&unscoped_q=mergejson). @hirofumi0810 and @simpleoier, could you comment this PR? Could you manage to reflect it to your local mergejson?

**Comment by simpleoier** [comment]
I'm good with this. I've updated my part to use it.

**Pull request merged**